### ISSUE TITLE
[LiquidCortex.jl] GH#57 — Reservoir hyperparameters are module consts and there is no RNG seed — parameter sweeps and reproducibility are both impossible

### ISSUE THREAD

**Issue by rmems** [rmems/LiquidCortex.jl#57]
## Summary

22 of 26 tunable quantities in `src/sparse_brain.jl` are module-level `const`s, the spectral radius is a bare local literal, and there is no RNG seed anywhere in the package. The single most common experiment in reservoir computing — sweep ρ × N and measure memory capacity — cannot be run, and no result the package produces is reproducible.

#5 made `n_in`/`n_out` configurable. This is the rest of that job.

## What a user can configure today

| Constructor | Tunable |
|---|---|
| `SparseBrain(tau_m; n_in, n_out, name)` | membrane τ, input dim, output dim, and a `name` that is never stored in the struct |
| `EnsembleBrain(; n_in, n_out)` | input dim, output dim |

## What is locked in source

| Line | Const | Value | Why it matters |
|---|---|---|---|
| `:27` | `N` | `65_536` | *The* reservoir hyperparameter; also sets the VRAM floor |
| `:28` | `CONN_PROB` | `0.01` | Sparsity ↔ memory-capacity tradeoff |
| `:191` | `target_rho` (a **local**, not even a const) | `0.9f0` | Spectral radius — the most-swept parameter in RC |
| `:29` | `DT` | `1.0f0` | Euler–Maruyama step |
| `:30` | `HIST_DEPTH` | `1000` | Fixes a 262 MB/lobe allocation |
| `:31` | `COV_SUBSAMPLE` | `8192` | Sets a 268 MB covariance return; `BoundsError` if `N` < 8192 |
| `:34-40` | `LOBE_TAUS`, `LOBE_NAMES`, `N_LOBES`, `LOBE_WEIGHTS` | | The entire multi-timescale premise |
| `:45-49` | `V_REST`, `V_THRESH`, `V_RESET`, `SIGMA`, `REFRAC_T` | | LIF biophysics + noise amplitude |
| `:56-57` | `TAU_TRACE`, `W_MAX` | | Primary STDP knobs |
| `:63-64` | `INHIBITION_GAIN`, `MAX_INHIBITION` | | Define the meaning of the public `inhibition` kwarg |

There is no legal override. `@eval LiquidCortex N = 4096` fails on a `const`, and would in any case leave ~40 stale inlined uses inconsistent.

## No RNG seed

`grep -rn "seed\|rng\|AbstractRNG\|MersenneTwister\|Xoshiro" src/` returns **nothing**. Every stochastic call uses the implicit global RNG: topology (`:176-178`), weight init (`:206, :211` → `cpu_randn_cu` at `:72`), per-tick noise (`:370/373`), and `randperm(N)` in `compute_reservoir_covariance!` (`:519`).

`Random.seed!(k)` before construction *does* make the default single-task path reproducible, so this is not total — but it is fragile enough to count as real reproducibility debt:

1. **No encapsulation.** One unrelated `rand()` between `seed!` and construction — a shuffle, a train/test split — silently changes the reservoir. Reproducing a published result means reproducing the caller's entire RNG call sequence.
2. **Two RNGs, two APIs.** `use_device_noise=true` uses CUDA.jl's device RNG, which `Random.seed!` does **not** seed (that needs `CUDA.seed!`). Nothing documents this; the `step!` docstring presents the flag as a performance switch.
3. **Silent path-dependence.** The `try/catch` at `:362-371` falls back from device to host RNG on any non-Interrupt/non-CUDA failure, with no warning. One transient hiccup permanently desynchronizes the stream from a reference run.
4. **Task-locality.** `default_rng()` is a `TaskLocalRNG`, so any `@async` in the user's harness gets a different stream.

Consecutive `compute_reservoir_covariance!` calls also subsample different neurons, so two covariance matrices from one brain are not comparable.

## Ecosystem comparison

Every peer library takes size as a runtime argument, and four of six seed explicitly. Verified against pinned commits / raw source:

| | size | connectivity | RNG seed |
|---|---|---|---|
| **LiquidCortex.jl** | **module const** | **module const** | **none** |
| ReservoirComputing.jl | positional arg | `sparsity=` kwarg | `rng` 1st positional, default `Xoshiro(1234)` |
| Brian2 | positional arg | `connect(p=)` | `brian2.seed()` |
| Nengo | positional arg | explicit indices | `seed=` on object, network *and* simulator |
| NEST | `Create(n=)` | `pairwise_bernoulli` | `nest.rng_seed` |
| BindsNET | `Nodes(n=)` | — | caller's job |
| snnTorch | via `nn.Linear` | — | caller's job |

No comparable Julia package uses a module `const` as its sole network size. The Julia Style Guide is explicit: *"functions should take arguments, instead of operating directly on global variables (aside from constants like `pi`)"* — a reservoir size is a problem dimension, not a universal constant.

## Proposed fix

A config struct carried as a `SparseBrain` field, with the current `const`s as defaults so nothing breaks:

```julia
@kwdef struct BrainConfig
    N::Int = 65_536;  conn_prob::Float64 = 0.01;  spectral_radius::Float32 = 0.9f0
    dt::Float32 = 1f0;  hist_depth::Int = 1000;   cov_subsample::Int = 8192
    v_rest::Float32 = -65f0; v_thresh::Float32 = -50f0; v_reset::Float32 = -70f0
    sigma::Float32 = 2f0; refrac_t::Int = 5; tau_trace::Float32 = 20f0
    w_max::Float32 = 1f0; inhibition_gain::Float32 = 15f0; max_inhibition::Float32 = 3f0
    rng::AbstractRNG = Random.default_rng()
end

SparseBrain(tau_m::Real; cfg::BrainConfig = BrainConfig(), n_in=14, n_out=16)
EnsembleBrain(; taus = LOBE_TAUS, weights = LOBE_WEIGHTS, cfg, n_in, n_out)
```

`_step_impl!` then reads `brain.cfg.*` (~15 substitution sites). Thread the stored `rng` through every `rand`/`randn`/`randperm` including the noise path, and select the noise source from the RNG's type rather than a boolean.

Two guards to add with it: clamp `cov_subsample` to `min(COV_SUBSAMPLE, N)` (currently `randperm(N)[1:8192]` throws for `N < 8192`), and note that `EnsembleBrain`'s machinery is already generic over `eb.lobes`/`eb.weights` — only the kwarg constructor is hardcoded to 4, so the lobe count is nearly free.

## Why this is the highest-leverage change in the package

It is a prerequisite for three other things:

- **Testing the core on CPU CI.** A 256-neuron brain runs the identical code path in milliseconds. Today `_step_impl!` executes zero times on any hosted runner.
- **A fast regression test for the spectral-radius fix.**
- **Any small-scale use at all.** The only exported reservoir is fixed at 65,536 neurons; the 2,048-neuron reference LSM exists but is unexported (#23 was closed, but `run_lsm_step` is still absent from the export list at `src/LiquidCortex.jl:128-130`).

## Scope boundaries

This does not by itself fix the spectral-radius estimator, add serialization, or add a `reset!` — all separate. It also does not make the package usable without a GPU: the constructor still hard-fails on a CPU-only machine, and does so only *after* ~40 s and ~1 GB of host allocation, because it never checks `_cuda_available[]`.

## Relationships

- Parent: none
- **Blocked by (native):** #56 (now closed by merged #65) — constructor overlap; `BrainConfig` would rewrite the same `SparseBrain` guards that #65 just landed. Re-check the native link after the closer settles; implementation can proceed once this is unblocked.
- **Blocks (native):** #55 — CPU-side core tests need configurable `N`
- **Related:** persist leftover from #60 (`save`/`load` needs config+seed; #70 closer is reset/free only), #53 (spectral radius should become a config field), #62 (reference-LSM export / small-scale path), #5 (closed; `n_in`/`n_out`)

## Provenance

Twelve-agent tech-debt audit, including a peer-library comparison verified against pinned commits. Full report: https://claude.ai/code/artifact/4978b7fa-8f0d-4b00-a03c-dc9a2acbc594

**Comment by rmems** [comment]
**Cursor agent** — Hygiene: native `blocked-by` #56 is now set because `BrainConfig` would rewrite the same `SparseBrain` constructor that open PR #65 is changing. Persist/`save`/`load` leftover from #60 stays here — do not close this issue when #70 merges (`Closes #60` is reset/free only). #55 remains blocked by this issue.
